### PR TITLE
Instance refactor

### PR DESCRIPTION
**Pull request by zach2good**
Oh boy, here it is. I've been trying to get this from proof of concept to usable refactor for _ages_.

**TODO**
- Update docs headers everywhere.
- A little more work to clean up Ambuscade
- Audit and make sure entry and gameplay of all existing "working" instances is still OK
- Kill all naked nested Lua access in luautils: `auto onInstanceZoneIn = lua["xi"]["zones"][name]["Zone"]["onInstanceZoneIn"];`

**Review Guide**
- `instance.lua` global updated to start looking like `bcnm.lua` - intended to be a wrapper around instance requirements, registration, queueing, entry etc.
- Re-indexed `instance_list` and laid out rules for ordering of that table, updated all references in code to the new indexes
- Updated `instance_entities` to use new indexes
- Filled in instance zone boilerplate for all instance zones (`onInstanceZoneIn` and `onInstanceLoadFailed`)
- `instance.lua` command to take you to instances
- Increasing the size of instanceId everywhere from uint8 to uint16 (tinyint to smallint)
- Mobs and Instance Mobs now properly populate packetName from the database. This doesn't change anything _yet_, but could eventually lead to us renaming mobs on the fly without using Renamer (TODO: Research into whats needed for this, the system works for Trusts, but not mobs)
- Add `IsValidInstanceID` to validate instance requests before we action them.
- `OnInstanceCreated` -> `OnInstanceCreatedCallback`, intended to wrap up the "here's some logic for the person going to the instance, just before we teleport them"
- Fixing the Puppet/Pet crash in instances by being more careful with their removal from the pet list.
- Fixing the "GM logs into an instance zone, and the zone spins up an instance, but it isn't really a proper instance, so the whole system freaks out" bug.
- Adding `TracyZoneScoped` to various misc places
- Starting a basic Ambuscade implementation. I observed that not only the mobs (in the same ID range) are swapped out every month, but the MOVES too. So I set up my Meeble and he has the Gigas name, and gigas moves. TODO: Find a way to approximate previous months Ambus without the need for windower/ashita intervention
- `Un-indented mob_family_system.sql`
- Added Meeble Family, no big deal.
- Added first Meeble to the game, no big deal.
- Added i119 Bozzetto Breadwinner who will kick your ass. Now the strongest mob in the game (name and animations broken, otherwise works great)

**Sanity Checklist**
- Repeated entry/exit
  - !instance 30000 (entry)
  - !instance 30000 (exit)
- Log in after crash
  - !instance 30000
  - Kill server
  - Log back in
  - Where do you land?
- Request invalid instance
- Force fail an instance while you're engaged with a mob (still black screen, like everything else :( )

**My old PRs for this:**
- https://github.com/topaz-next/topaz/pull/2472
- https://github.com/topaz-next/topaz/pull/2547

**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [x] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

### CODE REVIEW

File: scripts/globals/instance.lua [hunk 1]
```diff
@@ -1,10 +1,308 @@
...
+    player:startEvent(csid, unpack(zoneLookup[1][2]))
+end
+
+xi.instance.onEventUpdate = function(player, csid, option)
+end
+
+xi.instance.onEventFinish = function(player, csid, option)
+end
+
+local function setInstanceLastTimeUpdateMessage(instance, players, remainingTimeLimit, texttable)
```
Comment (TeoTwawki) [inline scripts/globals/instance.lua]: texttable vs textTable?

Comment (zach2good) [reply scripts/globals/instance.lua]: Yeah, that was just copy/pasting what was already there - I'll probably turn it into `text` when I come back around to clean

File: scripts/globals/ambuscade.lua [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+end
+
+-----------------------------------
+-- Ambuscade Tome
+-----------------------------------
+xi.ambuscade.onTradeTome = function(player, npc, trade)
+end
+
+xi.ambuscade.onTriggerTome = function(player, npc)
+    local currentPage = 735
```
Comment (claywar) [inline scripts/globals/ambuscade.lua:108]: Unused as well

File: scripts/globals/ambuscade.lua [hunk 3]
```diff
@@ -0,0 +1,75 @@
+-----------------------------------
+-- Ambuscade
+-----------------------------------
+-- Ambuscade_Tome : !pos -28.030 -15.500 52.279 249
+-- Gorpa-Masorpa  : !pos -27.584 -15.990 52.565 249
+-----------------------------------
+require("scripts/globals/utils")
+local mhauraID = require("scripts/zones/Mhaura/IDs")
```
Comment (claywar) [inline scripts/globals/ambuscade.lua]: Just some CI complaints on those two zone requires being unused

**Review by claywar — APPROVED** [review]
Looks good to me, might want to comment out those unused vars in ambuscade.lua if they're going to be used in the future 👍

**Comment by zach2good** [comment]
Not much left TODO here, will finish up this week

**Comment by zach2good** [comment]
Just caught up with Teo's resistance changes, things _should_ be working and nothing got trampled.

Still todo:
- Audit and make sure entry and gameplay of all existing "working" instances is still OK
- Kill all naked nested Lua access in luautils: auto onInstanceZoneIn = lua["xi"]["zones"][name]["Zone"]["onInstanceZoneIn"];
- I finally noticed that the entrance "animation" when you enter an instance is actually a CS and not an instanceEntry packet. So I need to rewrite some things in the global to handle that

**Comment by Era-Lusiphur** [comment]
>     * I finally noticed that the entrance "animation" when you enter an instance is actually a CS and not an instanceEntry packet. So I need to rewrite some things in the global to handle that

Correctamundo. Most of them are fairly nice looking, too.

**Comment by TeoTwawki** [comment]
>     * I finally noticed that the entrance "animation" when you enter an instance is actually a CS and not an instanceEntry packet. So I need to rewrite some things in the global to handle that

So like I was thinking during your stream then, need daisy chain the scenes together.

**Comment by zach2good** [comment]
OK, it's done and ready for final review. The only thing left to test on my end is party entry/rejection inside `xi.instance.onEventUpdate`. I have tested that returning false out of the function bails out the entire entry process, so it shouldn't need much work/testing

**Comment by zach2good** [comment]
The "raw table access" cleanup can come after, I just want to get this in. I've looked at all the remaining TODOs and they should be OK

**Comment by zach2good** [comment]
Tested with parties, should be good to go!

**Review by claywar — APPROVED** [review]

**Review by Era-Lusiphur — APPROVED** [review]
Tested earlier today, ran into some crashes when going into instances repeatedly. Discussed with zach directly. This last commit d5576e1 appears to have resolved those issues. Everything that I threw at it appeared to work. LGTM!

**Comment by zach2good** [comment]
image

**Pull request merged**